File: docs/js/ingest.js
// Decode a .txt of unknown encoding, then normalise it.
//
// Browsers ship WHATWG decoders for gb18030, big5, shift_jis and euc-kr, so the
// hard part is choosing between them. Same approach as the Python version:
// decode with each candidate and score by how *ordinary* the characters are.
// Real Chinese leans on a few hundred common characters; a wrong CJK codec
// produces rare-ideograph soup.

const COMMON_SIMPLIFIED =
  '的一是了我不人在他有这个上们来到时大地为子中你说生国年着就那和要她出也得里后自以' +
  '会家可下而过天去能对小多然于心学么之都好看起发当没成只如事把还用第样道想作种开美' +
  '总从无情己面最女但现前些所同日手又行意动方期它头经长儿回位分爱老因很给名法间知世' +
  '什两次使身者被高已亲其进此话常与活正感';
const COMMON_TRADITIONAL =
  '們這來時個國學會後說對於發過麼萬點爲產經體實現關樣頭長無愛親車東馬語話買賣讀寫聽' +
  '見開門問題還與從當隻總處務準辦種義許將軍書氣聲區內兩戰師應該沒覺得裡邊變傅劍靈魔' +
  '勢陣訣煉輩隨術龍鳳識脈遠處';
const COMMON = new Set([...COMMON_SIMPLIFIED, ...COMMON_TRADITIONAL]);

const TRAD_ONLY = new Set('們這來時國學會後說對發過麼萬點產經體實現關樣頭長無愛親車東馬語話買賣讀寫聽見開門問題還與從個');
const SIMP_ONLY = new Set('们这来时国学会后说对发过么万点产经体实现关样头长无爱亲车东马语话买卖读写听见开门问题还与从个');

// Ordered by coverage. gb18030 decodes almost anything, so it never wins on
// decodability alone — only on score.
const CANDIDATES = [
  'utf-8',
  'gb18030',
  'big5',
  'utf-16le',
  'utf-16be',
  'shift_jis',
  'euc-kr',
  'windows-1252',
];

const SAMPLE_BYTES = 400_000;

const isHan = (ch) => ch >= '一' && ch <= '鿿';

function score(text) {
  const n = text.length;
  if (!n) return -Infinity;
  let cjk = 0;
  let common = 0;
  let bad = 0;
  let asciiOk = 0;
  for (const ch of text) {
    if (isHan(ch)) {
      cjk++;
      if (COMMON.has(ch)) common++;
    } else if (COMMON.has(ch)) {
      common++;
    }
    if (ch === '\n' || ch === '\r' || ch === '\t') {
      asciiOk++;
      continue;
    }
    const code = ch.codePointAt(0);
    // C0/C1 controls, replacement char, private use, surrogates.
    if (code < 0x20 || (code >= 0x7f && code <= 0x9f) || code === 0xfffd ||
        (code >= 0xe000 && code <= 0xf8ff)) {
      bad++;
    } else if (ch >= ' ' && ch <= '~') {
      asciiOk++;
    }
  }
  let s = -60 * (bad / n);
  if (cjk / n > 0.03) s += 100 * (common / cjk);
  else s += 45 * (asciiOk / n);
  return s;
}

/**
 * Decode a prefix, tolerating a character split by the cut.
 *
 * Truncating a byte sample almost always lands mid-character in a multibyte
 * encoding. A non-streaming fatal decode then throws, the *correct* codec is
 * discarded, and scoring is left choosing between codecs that can never fail
 * (windows-1252 decodes anything) — so Chinese text ends up as mojibake.
 * `{stream: true}` holds the dangling tail instead of throwing, while a codec
 * that genuinely cannot read the content still throws.
 */
function decodeSample(bytes, label) {
  try {
    return new TextDecoder(label, { fatal: true }).decode(
      bytes.subarray(0, SAMPLE_BYTES),
      { stream: true },
    );
  } catch {
    return null;
  }
}

export function scoreCandidates(bytes) {
  return CANDIDATES.map((label) => {
    const text = decodeSample(bytes, label);
    return text === null
      ? { encoding: label, score: -Infinity, sample: '<cannot decode this content>' }
      : { encoding: label, score: Math.round(score(text) * 100) / 100,
          sample: text.slice(0, 120).replace(/\n/g, '⏎') };
  }).sort((a, b) => b.score - a.score);
}

const BOMS = [
  [[0xef, 0xbb, 0xbf], 'utf-8'],
  [[0xff, 0xfe, 0x00, 0x00], 'utf-16le'],
  [[0xff, 0xfe], 'utf-16le'],
  [[0xfe, 0xff], 'utf-16be'],
];

function startsWith(bytes, prefix) {
  return prefix.every((b, i) => bytes[i] === b);
}

export function decode(bytes) {
  for (const [prefix, label] of BOMS) {
    if (startsWith(bytes, prefix)) {
      return { text: new TextDecoder(label).decode(bytes), encoding: label + ' (BOM)' };
    }
  }

  // A clean strict UTF-8 decode is decisive: no other codec produces valid
  // UTF-8 multibyte sequences by accident at any length.
  try {
    return { text: new TextDecoder('utf-8', { fatal: true }).decode(bytes), encoding: 'utf-8' };
  } catch { /* not UTF-8 */ }

  // BOM-less UTF-16 is unmistakable from NUL placement.
  const head = bytes.subarray(0, 8000);
  let nul = 0;
  let even = 0;
  let odd = 0;
  for (let i = 0; i < head.length; i++) {
    if (head[i] === 0) {
      nul++;
      if (i % 2 === 0) even++;
      else odd++;
    }
  }
  if (nul > head.length * 0.2) {
    const label = even > odd ? 'utf-16be' : 'utf-16le';
    return { text: new TextDecoder(label).decode(bytes), encoding: label + ' (no BOM)' };
  }

  let best = null;
  for (const label of CANDIDATES) {
    const text = decodeSample(bytes, label);
    if (text === null) continue;
    const s = score(text);
    if (!best || s > best.score) best = { score: s, label };
  }
  if (!best) return { text: new TextDecoder('utf-8').decode(bytes), encoding: 'utf-8 (with errors)' };
  return { text: new TextDecoder(best.label).decode(bytes), encoding: best.label };
}

export function detectLanguage(text) {
  const sample = text.slice(0, 200_000);
  let han = 0;
  let kana = 0;
  let hangul = 0;
  for (const ch of sample) {
    if (isHan(ch)) han++;
    else if (ch >= '぀' && ch <= 'ヿ') kana++;
    else if (ch >= '가' && ch <= '힣') hangul++;
  }
  const total = Math.max(sample.length, 1);
  if (hangul / total > 0.05) return 'ko';
  if (kana / total > 0.02) return 'ja';
  if (han / total > 0.05) {
    let trad = 0;
    let simp = 0;
    for (const ch of sample) {
      if (TRAD_ONLY.has(ch)) trad++;
      else if (SIMP_ONLY.has(ch)) simp++;
    }
    return trad > simp * 1.2 ? 'zh-Hant' : 'zh-Hans';
  }
  return 'en';
}

export function normalise(text) {
  let out = text.replace(/\r\n?/g, '\n').replace(/﻿/g, '');
  // Strip control characters the reader can't use, keeping \n and \t.
  out = out.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\x9F]/g, '');
  out = out
    .split('\n')
    // Leading full-width spaces are the source file's indentation; the
    // stylesheet owns indentation in the EPUB.
    .map((line) => line.replace(/^[　\s]+/, '').replace(/\s+$/, ''))
    .join('\n');
  return out.replace(/\n{3,}/g, '\n\n').replace(/^\n+|\n+$/g, '');
}

export function load(bytes) {
  const { text: raw, encoding } = decode(bytes);
  const text = normalise(raw);
  return {
    text,
    lines: text.split('\n'),
    encoding,
    language: detectLanguage(text),
    charCount: text.length,
  };
}

File: txt2epub/ingest.py
"""Decode a .txt file of unknown encoding and normalise it into clean text.

Chinese plain-text ebooks in the wild are GB18030, Big5 or UTF-8, usually with
no BOM and often mislabelled. Rather than trusting a single detector we decode
with every plausible codec and score the results: genuine Chinese text draws
heavily on a few hundred very common characters, while a wrong CJK codec
produces a soup of rare ideographs. That ratio separates them cleanly.
"""

from __future__ import annotations

import codecs
import re
import unicodedata
from dataclasses import dataclass

# The most frequent characters in simplified and traditional Chinese. Real text
# covers a large fraction of this set; mojibake essentially never does.
_COMMON_SIMPLIFIED = (
    "的一是了我不人在他有这个上们来到时大地为子中你说生国年着就那和要她出也得里后自以"
    "会家可下而过天去能对小多然于心学么之都好看起发当没成只如事把还用第样道想作种开美"
    "总从无情己面最女但现前些所同日手又行意动方期它头经长儿回位分爱老因很给名法间知世"
    "什两次使身者被高已亲其进此话常与活正感"
)
_COMMON_TRADITIONAL = (
    "們這來時個國學會後說對於發過麼萬點爲產經體實現關樣頭長無愛親車東馬語話買賣讀寫聽"
    "見開門問題還與從當隻總處務準辦種義許將軍書氣聲區區內兩戰師實應該還沒麼樣覺得裡邊"
    "說話當然變樣師傅劍靈魔氣勢陣訣煉輩隨憐術術術龍鳳靈識靈氣體內經脈遠遠遠處聽見來說"
)
_COMMON = set(_COMMON_SIMPLIFIED + _COMMON_TRADITIONAL)

_TRAD_ONLY = set("們這來時國學會後說對發過麼萬點產經體實現關樣頭長無愛親車東馬語話買賣讀寫聽見開門問題還與從個")
_SIMP_ONLY = set("们这来时国学会后说对发过么万点产经体实现关样头长无爱亲车东马语话买卖读写听见开门问题还与从个")

_BOMS = (
    (b"\xef\xbb\xbf", "utf-8-sig"),
    (b"\xff\xfe\x00\x00", "utf-32-le"),
    (b"\x00\x00\xfe\xff", "utf-32-be"),
    (b"\xff\xfe", "utf-16-le"),
    (b"\xfe\xff", "utf-16-be"),
)

# Ordered by how much of the world's .txt they cover. gb18030 will "succeed" on
# almost any byte string, so it never wins on decodability alone — only on score.
# The BOM-less UTF-16 variants matter: Windows Notepad's "Unicode" save produces
# them, and every 8-bit codec turns them into NUL-riddled garbage.
_CANDIDATES = (
    "utf-8",
    "gb18030",
    "big5hkscs",
    "utf-16-le",
    "utf-16-be",
    "shift_jis",
    "euc-kr",
    "cp1252",
    "latin-1",
)


@dataclass
class Document:
    text: str
    lines: list[str]
    encoding: str
    language: str  # BCP-47: zh-Hans, zh-Hant, ja, ko, en
    char_count: int

    @property
    def is_cjk(self) -> bool:
        return self.language.startswith(("zh", "ja", "ko"))


def _score(text: str) -> float:
    """Higher is more likely to be a correct decoding."""
    n = len(text)
    if n == 0:
        return -1e9
    cjk = 0
    common = 0
    bad = 0
    ascii_ok = 0
    for ch in text:
        if "一" <= ch <= "鿿":
            cjk += 1
            if ch in _COMMON:
                common += 1
        elif ch in _COMMON:
            common += 1
        if ch in "\n\r\t":
            ascii_ok += 1
            continue
        cat = unicodedata.category(ch)
        if cat in ("Cc", "Cf", "Co", "Cn", "Cs"):
            bad += 1
        elif " " <= ch <= "~":
            ascii_ok += 1

    score = -60.0 * (bad / n)
    if cjk / n > 0.03:
        # Chinese/Japanese text: judge by how ordinary the characters are.
        score += 100.0 * (common / cjk)
    else:
        score += 45.0 * (ascii_ok / n)
    return score


def decode(data: bytes) -> tuple[str, str]:
    """Return (text, encoding-name) for a bytes blob of unknown encoding."""
    for bom, enc in _BOMS:
        if data.startswith(bom):
            return data.decode(enc, errors="replace"), enc

    # A clean strict UTF-8 decode is decisive — no other codec produces valid
    # UTF-8 multibyte sequences by accident at any length.
    try:
        return data.decode("utf-8"), "utf-8"
    except UnicodeDecodeError:
        pass

    # BOM-less UTF-16 is unmistakable from NUL placement, and scoring can't see
    # it because an odd-length sample makes both variants decode to noise.
    head = data[:8000]
    nul = head.count(0)
    if nul > len(head) * 0.2:
        even = sum(1 for i in range(0, len(head) - 1, 2) if head[i] == 0)
        odd = sum(1 for i in range(1, len(head), 2) if head[i] == 0)
        enc = "utf-16-be" if even > odd else "utf-16-le"
        return data.decode(enc, errors="replace"), enc + " (no BOM)"

    best: tuple[float, str, str] | None = None
    for enc in _CANDIDATES:
        text = _decode_sample(data, enc)
        if text is None:
            continue
        s = _score(text)
        if best is None or s > best[0]:
            best = (s, enc, text)

    if best is None:
        return data.decode("utf-8", errors="replace"), "utf-8 (with errors)"

    _, enc, _ = best
    return data.decode(enc, errors="replace"), enc


_SAMPLE_BYTES = 400_000


def _decode_sample(data: bytes, enc: str) -> str | None:
    """Decode a prefix of `data`, tolerating a character split by the cut.

    Truncating a byte sample almost always lands mid-character in a multibyte
    encoding. A plain strict `bytes.decode` then raises, the *correct* codec is
    discarded, and scoring is left to choose between codecs that can never
    fail — cp1252 and latin-1 happily decode anything, so Chinese text ends up
    as mojibake. An incremental decoder with `final=False` holds the dangling
    tail bytes instead of raising, while still rejecting a codec that genuinely
    cannot read the content.
    """
    try:
        decoder = codecs.getincrementaldecoder(enc)(errors="strict")
        return decoder.decode(data[:_SAMPLE_BYTES], final=False)
    except (UnicodeDecodeError, LookupError):
        return None


def score_candidates(data: bytes) -> list[tuple[str, float, str]]:
    """Every codec's score and a text sample — for the diagnostics endpoint."""
    out = []
    for enc in _CANDIDATES:
        text = _decode_sample(data, enc)
        if text is None:
            out.append((enc, float("-inf"), "<cannot decode this content>"))
            continue
        out.append((enc, round(_score(text), 2), text[:120].replace("\n", "⏎")))
    return sorted(out, key=lambda row: -row[1])


def detect_language(text: str) -> str:
    sample = text[:200_000]
    han = sum(1 for c in sample if "一" <= c <= "鿿")
    kana = sum(1 for c in sample if "぀" <= c <= "ヿ")
    hangul = sum(1 for c in sample if "가" <= c <= "힣")
    total = max(len(sample), 1)

    if hangul / total > 0.05:
        return "ko"
    if kana / total > 0.02:
        return "ja"
    if han / total > 0.05:
        trad = sum(1 for c in sample if c in _TRAD_ONLY)
        simp = sum(1 for c in sample if c in _SIMP_ONLY)
        return "zh-Hant" if trad > simp * 1.2 else "zh-Hans"
    return "en"


_CONTROL_OK = {"\n", "\t"}


def normalise(text: str) -> str:
    text = text.replace("\r\n", "\n").replace("\r", "\n").replace("﻿", "")
    text = "".join(
        ch for ch in text if ch in _CONTROL_OK or unicodedata.category(ch)[0] != "C"
    )
    out = []
    for line in text.split("\n"):
        # Leading full-width spaces are the source file's indentation; the
        # stylesheet owns indentation in the EPUB, so strip them here.
        out.append(line.strip("　  \t"))
    text = "\n".join(out)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip("\n")


def load(data: bytes) -> Document:
    raw, encoding = decode(data)
    text = normalise(raw)
    return Document(
        text=text,
        lines=text.split("\n"),
        encoding=encoding,
        language=detect_language(text),
        char_count=len(text),
    )
